FLOE-430: Storing preferences to the Preferences Server

jobara opened the pull request:
Connecting the FD Server to the Preferences Server via the Oauth2 security layer.

https://issues.fluidproject.org/browse/FLOE-430
https://issues.gpii.net/browse/GPII-1337

This pull request includes changes from https://github.com/GPII/first-discovery-server/pull/2

jobara committed:
FLOE-430: In progress commit, unit tests failing.

jobara committed:
FLOE-430: Unit tests passing again

Missed changing a require statement.

jobara committed:
FLOE-430: Moved kettle to a full dependency

jobara committed:
FLOE-430: Roughing in view query parameter

The view query parameter on the /user endpoint is used to specify which ontology the prefences are saved in.

jobara committed:
Merge branch 'FLOE-423' into FLOE-430

jobara committed:
FLOE-430: Updated to support debugging

jobara committed:
FLOE-430: Adding JSON body parsing

jobara committed:
FLOE-430: Connected to security server

Needs a lot of refactoring and tests. This is just the first end-to-end working implementation.

jobara committed:
FLOE-430: Missed committing dataSource

jobara committed:
Merge branch 'FLOE-423' into FLOE-430

jobara committed:
FLOE-430: Removed dataSource file

There didn't seem to be enough reusable modifications to justify a new dataSource grade.

jobara committed:
FLOE-430: refactored error handling

jobara committed:
FLOE-430: Refactoring of request handler

jobara committed:
FLOE-430: Left a TODO about tests

jobara committed:
Merge branch 'FLOE-423' into FLOE-430

jobara committed:
FLOE-430: Using Nock to mock security requests

The tests are all passing, but need to be refactored.

jobara committed:
FLOE-430: Refactoring - reordering functions

jobara committed:
FLOE-430: Added a TODO about security config

jobara committed:
FLOE-430: Updated kettle and infusion dependencies

jobara committed:
FLOE-430: Breaking down datasource urls

jobara committed:
FLOE-430: Added a configuration loader

Added config.js which can load in configuration from a config file or environment variables.
This has not yet been hooked up to the server.

jobara committed:
FLOE-430: Wired up the prefs server config

jobara committed:
FLOE-430: Fixed tests

jobara committed:
FLOE-430: Added tests for config.js

jobara committed:
FLOE-430: renamed test

jobara committed:
FLOE-430: Updating port # in readme

jobara committed:
Merge branch 'FLOE-423' into FLOE-430

jobara commented:
@amb26 could you please review this pull request?

jobara committed:
FLOE-430: Renaming ENV for path variables

jobara committed:
FLOE-430: Updated readme

jobara committed:
FLOE-430: Fixed some urls

jobara committed:
FLOE-430: Moved configuration section into usage

amb26 commented on src/js/config.js:
```diff
@@ -0,0 +1,127 @@
+ *                          configuration object.
+ * @returns {Object} - the configuration object
+ */
+gpii.firstDiscovery.server.config.getENV = function (ENVMap) {
```
I think capitalisation of "ENV" can be "Env" here and throughout this implementation - it isn't any kind of acronym, after all

amb26 commented on src/js/config.js:
```diff
@@ -0,0 +1,127 @@
+ * @param schema {Object} - a JSON schema document based on AJV (see: https://github.com/epoberezkin/ajv)
+ * @returns {Object} - the combined and validated configuraiton object
+ */
+gpii.firstDiscovery.server.config.getConfig = function (file, ENVMap, schema) {
```
This scheme isn't very extensible - and will make it increasingly hard to aggregate multiple sources of integration within config files in a declarative way. It suffers from the same problems of "function-ism" that we identified before in our prefs server driver - as soon as a configuration point consists of a function rather than a grade, we create a point of fragility which keeps propagating.
What I suggest instead is that you allow the user to just write standard declarative configs, with interpolation possible from the environment using a syntax such as "${environmentVariables}.GPII_OAUTH2_AUTH_GRANT_TYPE" etc.

A suggested scheme for this is to manually create a global component on startup which iterates over process.env and assigns each of the values found there into a top-level member.

I believe you're using a version of Infusion new enough to use the new "root" grade "fluid.resolveRootSingle" - http://docs.fluidproject.org/infusion/development/Contexts.html#global-components-fluid.resolveRoot-and-fluid.resolveRootSingle

By using this API you can ensure that the component you create, say, "gpii.environmentVariables" is resolvable by the above name from all configs.

amb26 commented on index.js:
```diff
@@ -12,8 +12,16 @@ https://raw.githubusercontent.com/GPII/first-discovery-server/master/LICENSE.txt
+var path = require("path");
 require("./src/js/firstDiscoveryServer.js");
+require("./src/js/config.js");
+
```
This would be a good place to register your module according to the new module API pattern - 

fluid.module.register("first-discovery-server", __dirname, require)

https://github.com/fluid-project/infusion-docs/pull/77/files#diff-d7540666721f8644efd29f0d5fb29613R53

Although we may as well take the opportunity to namespace this module properly in npm since we plan to deploy everything there before long - I suggest a new package name of "gpii-first-discovery-server" but open to alternatives.

amb26 commented on index.js:
```diff
@@ -12,8 +12,16 @@ https://raw.githubusercontent.com/GPII/first-discovery-server/master/LICENSE.txt
 gpii.firstDiscovery.server({
-    port: process.env.FIRST_DISCOVERY_SERVER_TCP_PORT
+    port: process.env.FIRST_DISCOVERY_SERVER_TCP_PORT,
+    preferencesConfig: oauth2Config
```
After the changes above, and the revision to the environment variable resolution scheme, this can just be a standard declarative config held in the project and can be started according to the standard Kettle init scheme (from an env var, command line argument, etc.)

jobara committed:
Merge branch 'master' into FLOE-430

jobara committed:
FLOE-430: Renaming ENV to Env

amb26 commented on package.json:
```diff
@@ -32,16 +32,18 @@
   "dependencies": {
+    "ajv": "^1.4.9",
+    "dedupe-infusion": "1.0.0",
     "first-discovery": "git+https://github.com/GPII/first-discovery.git",
```
Use of raw git urls binding to a project master is hazardous and can allow a project to break without warning. Please bind against particular revisions of these projects instead

jobara replied:
Should we publish First Discovery, or just use the latest rev hash?

amb26 replied:
Either is fine - in general we should get into the habit of publishing more projects. We should adopt a general standard of all our npm artifacts being prefixed with gpii- , e.g. "gpii-first-discovery" in its package.json. if we decide to do this.

jobara replied:
I've filled a [JIRA](https://issues.gpii.net/browse/GPII-1519) and [pull request](https://github.com/GPII/first-discovery/pull/107) for adding the "gpii-" prefix to the first discovery tool.

amb26 commented on package.json:
```diff
@@ -32,16 +32,18 @@
-    "infusion": "2.0.0-dev.20151007T134312Z.8c33cd4",
-     "dedupe-infusion": "1.0.0"
+    "infusion": "git://github.com/amb26/infusion.git#d9ab2cea604cc3108bdf028d70874d339df22720",
+    "kettle": "git://github.com/amb26/kettle.git#KETTLE-32"
```
Similarly, binding against a branch target creates a risk here since the branch might be updated in a conflicting way

jobara replied:
do you have a recommendation on which revision to take for the branch, or i could just grab the last commit here.

jobara committed:
FLOE-430: Updated package.json file

1) added the gpii- prefix to the name
2) updated the version number
3) pulling in specific versions of dependencies

jobara committed:
FLOE-430: Switching to using Kettle Config

Making use of Kettle Configs to load and configure the FD Server

jobara committed:
FLOE-430: removed infusion-dedupe

jobara committed:
FLOE-430: Updated README

Updated the README file based on using Kettle configs to launch the app

jobara commented:
@amb26 ready for more code review. Switched to using kettle configs to launch the first discovery server. I've also updated the README. There may be a minor change to the docs coming to more specially mention how to use the secrets.yml file, once the changes to the vagrant/ansible role have been made.

jobara committed:
FLOE-430: fixed typo

amb26 commented on README.md:
```diff
@@ -4,16 +4,72 @@ A server side implementation to serve a [First Discovery Editor](https://github.
-node index.js
+# or using an environment variable to specify
+# the configName
+NODE_EVN=<configName> node node_modules/kettle/init.js <configPath>
```
Typo here EVN

amb26 commented on src/js/configUtils.js:
```diff
@@ -0,0 +1,95 @@
+ * Exposes the environment variables available to the node process, to a
+ * form that is accessible via the IoC system.
+ */
+fluid.defaults("gpii.firstDiscovery.environment", {
```
All of this stuff can just be put in plain gpii.\* rather than gpii.firstDiscovery.\* since it consists of general utilities that will be helpful across the project

amb26 commented on src/js/configUtils.js:
```diff
@@ -0,0 +1,95 @@
+        validationError: null
+    },
+    listeners: {
+        validationError: "gpii.firstDiscovery.schema.error",
```
These listeners should be namespaced as per our best practices

amb26 commented on src/js/configUtils.js:
```diff
@@ -0,0 +1,95 @@
+ * @throws {Error} - throws an error with the provide errorMsg
+ */
+gpii.firstDiscovery.schema.error = function (errorMsg) {
+    throw new Error(errorMsg);
```
The use of fluid.fail here is preferable since it is more open to customisation at the framework level

amb26 commented on src/config/environment.json:
```diff
@@ -0,0 +1,24 @@
+        "port": "{environment}.env.FIRST_DISCOVERY_SERVER_TCP_PORT",
+        "preferencesConfig": {
+            "securityServer": {
+                "port": "{environment}.env.GPII_OAUTH2_TCP_PORT",
```
I worry that "environment" is a bit too generic of a context name to use here -- what do you think about use of a grade name such as "gpii.envVars"? Or perhaps "gpii.resolvers.env" - preparing for a world in which we have a larger family of these things.

jobara replied:
Based on the channel discussion the IoC expression will be {gpii.resolvers.env}.vars.someName.
Note that the property containing the variables changes from "env" to "vars".

https://botbot.me/freenode/fluid-work/2015-12-14/?msg=56128841&page=1

amb26 commented on provisioning/vars.yml:
```diff
@@ -4,14 +4,11 @@
 nodejs_app_tcp_port: 8088
 
-nodejs_version: 0.10.40
+nodejs_version: 0.12.7
```
We may as well just move all the way up to our LTS version of 4.2.1 since I don't think we have any sensitive dependencies in this project.

jobara committed:
FLOE-430: Corrected typo

jobara committed:
FLOE-430: Renamed grades in configUtils

jobara committed:
FLOE-430: Updating VM to Node 4.2.1

jobara committed:
FLOE-430: Switched to fluid.fail

jobara committed:
FLOE-430: Change env member to vars

This makes it more generic for the future where we may have different types of
gpii.resolvers.

jobara committed:
FLOE-430: Fixing Unit tests

jobara committed:
FLOE-430: switched back to {environment}

Switched back to using the {environment} IoC reference as the singleRoot reference did not seem to be working.
Will attempt to put back to {gpii.resolvers.env} when it's working.

jobara commented:
@amb26 ready for more review. Had to use `{environment}` instead of `{gpii.resolvers.env}` in the environment.json config file as the singleRoot reference does not seem to be working.

jobara committed:
FLOE-430: Putting back manual infusion dedupe

Added back the manual Infusion dedupe, as it seems that gpii-express does not support the auto deduping.

jobara committed:
FLOE-430: Fixed paths and refactored

Fixed the paths to the gpii-first-discovery web app.

Also refactored how the environment variables are referenced using IoC.

jobara commented:
@amb26 ready for more review.

jobara committed:
FLOE-430: Updated README with info on secrets

Add information about using secret environment variables in the VM.

jobara commented:
@amb26 there is one more change that will need to be made before merging to support the secrets. Working on this with @avtar

jobara committed:
FLOE-430: Fixed typos

jobara commented:
@amb26 here's the JIRA for porting from gpii-express to kettle.
https://issues.gpii.net/browse/GPII-1538

jobara committed:
FLOE-430: Updated ansible config for secrets

Now secret environment variables can be added via the provisioning/secrets.yml file, which is not tracked by git.

jobara commented:
@amb26 this is ready for review and merge. The secrets should now be working.

jobara committed:
FLOE-430: Pointing at idi-ops repos

Updated the references from avtar's repo to the ones from idi-ops

jobara committed:
FLOE-430: Logging request errors

jobara committed:
FLOE-430: Code cleanup

jobara commented:
@amb26 added logging, ready for more review.

jobara committed:
FLOE-430: Updated demo url

jobara committed:
FLOE-430: Expanding error object
